File: credit.py
from cs50 import get_string
from math import floor

cn_str = get_string("Card Number: ")
n_digits = len(cn_str)

orig_cn = []
mult_numbers = []
cn_int = int(cn_str)
calc = 0

#copy into list (backwards)
for i in range(n_digits):

    orig_cn.append(cn_int % 10)
    cn_int = cn_int / 10

#to multiply every second value by 2, starting at second last value
#and store all new values in 2nd list 'mult_numbers'
for i in range(n_digits):

    x = i
    #if index is even
    if (x % 2 == 0):
        calc += orig_cn[i]
        print(f"{orig_cn[i]}")
    else:
        product = orig_cn[i] * 2
        #will this make product the modulus of product? probs not
        calc += product % 10 + floor(product / 10)
        #calc += orig_cn[i]*2 % 10 + orig_cn[i]*2 /10

print(f"calc: {calc}")
if calc % 10 != 0:
    print("INVALID")
    exit()

#which type of card
if n_digits==13 or n_digits==16 and orig_cn[n_digits - 1]==4:

        print("VISA")

elif n_digits==16 and orig_cn[15]==5 and 1 <= orig_cn <= 5:

        print("MASTERCARD")

elif n_digits==15 and orig_cn[14]==3 and orig_cn[13]==4 or orig_cn[13]==7:

        print("AMEX")

else:
        print("INVALID")

File: credit.c
#include <stdio.h>
#include <cs50.h>
#include <stdarg.h>

//introduction to ord_mag and add functions (see bottom)
int ord_mag(long x);
int main(void)
{
    //introducing sum, i, n_digits and c_number
int sum=0;
int i;
int n_digits;
long long c_number;
//Prompt user for number  
    c_number = get_long_long("Number\n");
    n_digits = ord_mag(c_number);

//introducing the first array orig_numbers
int orig_numbers[n_digits];
//To store the long long value as an array called orig_numbers
for (i=0; i < n_digits; i++)
{
    orig_numbers[i] = c_number%10;
    c_number /= 10;
}
//to introduce mult_numbers
int mult_numbers[n_digits];
//to multiply every second value by 2 and store all new values in 2nd array 'mult_numbers'
for(i=1; i<n_digits; i+=2)
{
    mult_numbers[i] = orig_numbers[i]*2;
    sum+= (mult_numbers[i] %10) +  (mult_numbers[i] /10);
}
//to use the original numbers and add each second value (starting from LAST) to sum
for(i=0;i<n_digits; i+=2)
{
    mult_numbers[i]=orig_numbers[i];
    sum+= (mult_numbers[i]%10)+(mult_numbers[i] /10);
}
//for validating a visa
if((n_digits==13|| n_digits==16) && sum%10==0 && orig_numbers[n_digits-1]==4)
{
    printf("VISA\n");
}
//to validate a mastercard
else if (n_digits==16 && orig_numbers[15]==5 && sum%10==0 && (orig_numbers[14]==1 || orig_numbers[14]==2 ||orig_numbers[14]==3 || orig_numbers[14]==4 || orig_numbers[14] ==5))
{
    printf("MASTERCARD\n");
}
//to validate an american express
else if(n_digits==15 && sum%10==0 && orig_numbers[14]==3 && (orig_numbers[13]==4 || orig_numbers[13]==7))
{
    printf("AMEX\n");
}
//otherwise print invalid - all other numbers bar 13,15 & 16 digit ones should have been weeded out by the first process of the ord_mag function
else 
{
    printf("INVALID\n");
}
}

//New outside function to check the number of digits
int ord_mag(long x)
{
    int n_digits = 0;
    for (long i=x; i>=1; n_digits++)
{
    i=i/10;
}
return n_digits;
}
